Vfb 318 Create Records in DB on submit

ebinezer04 opened the pull request:
## What's changed
Added a submit button at the bottom of the data grid component that submits the batch data into the database. 
On submit, a success or error dialog will display.
Added error checks for invalid/incomplete data, a further PR will correctly convey this information to the user.
Implement an add button to add new rows.
Implement a button on the navigation bar to switch to the batch creation page.


## Checklist
- [x] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [x] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test`

ebinezer04 committed:
VFB-318 WIP

VFB-318 WIP 2

ebinezer04 committed:
VFB-318 Implement submit functions for clients and parcels

ebinezer04 committed:
VFB-318 Fix Packing Slot UUID error

ebinezer04 committed:
VFB-318 Display success message on batch database addition

ebinezer04 committed:
VFB-318 Add success modal

ebinezer04 committed:
VFB-318 Implement audit log after submitting

ebinezer04 committed:
VFB-318 Rebase fixes

ebinezer04 committed:
VFB-318 Finish rebase fixes, lint and start empty cell checks

VFB-318 Remove duplicate Data Grid file

ebinezer04 committed:
VFB-318 Remove problematic Data Grid cypress test

VFB-318 Change function names and remove unnecessary console logs

VFB-318 Lint

VFB-318 Fix function argument error

rjesson-softwire commented on src/app/parcels/batch/displayComponents/BatchParcelDataGrid.tsx:
```diff
@@ -52,50 +59,173 @@ export const defaultTableState: BatchTableDataState = {
     );
+
+    const handleSubmit = async (): Promise<void> => {
+        console.log(tableState.batchDataRows);
```
are these console.logs supposed to be here?

rjesson-softwire commented on src/app/parcels/batch/submitTableData.ts:
```diff
@@ -0,0 +1,315 @@
+};
+
+const checkForRequiredClientData = (client: BatchClient): boolean => {
+    const hasRequiredFields: boolean =
+        client.fullName !== null &&
+        client.address !== null &&
+        client.adultInfo !== null &&
+        client.childrenInfo !== null &&
+        client.listType !== null &&
+        client.babyProducts !== null &&
+        client.deliveryInstructions !== null &&
+        client.extraInformation !== null &&
+        client.attentionFlag !== null &&
+        client.signpostingCall !== null &&
+        client.signpostingCall !== null &&
+        client.attentionFlag !== null;
+
+    return hasRequiredFields;
+};
+
+const checkForRequiredParcelData = (parcel: ParcelData): boolean => {
+    const hasRequiredFields: boolean =
+        parcel.packingDate !== null &&
+        parcel.packingSlot !== null &&
+        parcel.shippingMethod !== null;
+
+    const validCollectionOrDelivery: boolean =
+        parcel.shippingMethod !== "Collection" ||
+        (parcel.shippingMethod === "Collection" && parcel.collectionInfo !== null);
+
+    return hasRequiredFields && validCollectionOrDelivery;
+};
+
+const checkForEmptyParcel = (parcel: ParcelData): boolean => {
+    return Object.values(parcel).every((value) => value == null);
+};
+
+const checkForEmptyRow = (data: BatchEditData): boolean => {
+    return data === emptyBatchEditData;
+};
```
these functions could be renamed to be clearer in exactly what theyre checking - taking the last one as an example, maybe a name like checkRowIsNotEmpty might be more appropriate? the name checkForEmptyRow sounds to me like it's taking an array of rows and checking if it contains an empty row.

rjesson-softwire commented on src/app/parcels/batch/submitTableData.ts:
```diff
@@ -0,0 +1,315 @@
+        return { clientId: null, error: { type: "failedToInsertClientAndFamily", logId } };
+    }
+
+    console.log(`Row ${rowId}: Client ${client.fullName}, ${clientId} added successfully`);
```
should this console log be here?

rjesson-softwire reviewed (commented):
generally looks really good! just a few little comments

ebinezer04 committed:
VFB-318 Add navigation bar button for batch creation

VFB-318 Test data change

ebinezer04 committed:
VFB-318 Implement add button to add new row

ebinezer04 committed:
VFB-318 Add confirmation dialog before submitting

ebinezer04 committed:
VFB-318 Fix add new row bug when editing new cells

VFB-318 Fix batch cypress test for add row bug

VFB-318 Lint

VFB-318 Remove unnecessary write local state

VFB-318 Lint

ebinezer04 committed:
VFB-318 Remove isRowEmpty check

VFB-318 Lint

Stuart-Knowles-SW commented on src/app/parcels/batch/displayComponents/BatchParcelDataGrid.tsx:
```diff
@@ -37,65 +45,277 @@ export interface BatchGridDisplayRow {
+        },
+        {} as Record<string, number>
+    );
+};
```
I really like this function, but I think there's a couple of tweaks we can make.
Firstly, lets use `??` rather than `||`, I think it makes the intention a bit clearer.

Secondly, let's declare the initial state above the `errors.reduce` call, to improve type safety:
`const initialState: Record<string, number> = {};`. Although this looks basically the same, it's more type safe because `as` overrides type errors, whereas declaring the type of a `const` doesn't. So `{'foo': 'bar'} as Record<string, number>` doesn't throw a type error but `const initialState: Record<string, number> = { 'foo': 'bar' }` does. Although either works in this case, we prefer the second pattern because it's safer in general. Using `as` is usually a red flag so we should avoid it when we can.

Stuart-Knowles-SW commented on src/app/parcels/batch/displayComponents/BatchParcelDataGrid.tsx:
```diff
@@ -37,65 +45,277 @@ export interface BatchGridDisplayRow {
     const [tableState, dispatch] = useLocalStorage(batchParcelsReducer, defaultTableState);
+    const [isConfirmationDialogVisible, setIsConfirmationDialogVisible] = useState(false);
+    const [isSuccessDialogVisible, setIsSuccessDialogVisible] = useState(false);
+    const [dialogMessage, setDialogMessage] = useState("");
```
Can we combine these two into one state? i.e., `dialogMessage` is `string | null` and the dialog is open based on `dialogMessage !== null`

Stuart-Knowles-SW commented on src/app/parcels/batch/displayComponents/BatchParcelDataGrid.tsx:
```diff
@@ -37,65 +45,277 @@ export interface BatchGridDisplayRow {
+
+        setIsConfirmationDialogVisible(false);
+        setIsSuccessDialogVisible(true);
+    };
```
For these comments, rather than talk about the next PR, can you create a Jira ticket to do these changes then reference that ticket number in a todo here. Then we can delete the second/third comments. Something like:
```
        // VFB-XXX improve messaging on batch form submission
        if (submitErrors.length == 0) {
            setDialogMessage("Successfully submitted these rows");
            handleReset();
        } else {
            setDialogMessage("Errors encountered during submission.");
            handleUnsubmittedRows(submitErrors);
        }
```

Stuart-Knowles-SW commented on src/app/parcels/batch/displayComponents/BatchParcelDataGrid.tsx:
```diff
@@ -37,65 +45,277 @@ export interface BatchGridDisplayRow {
+
+            <Dialog
+                open={isConfirmationDialogVisible}
+                sx={{
+                    "& .MuiDialog-paper": {
+                        padding: "1rem",
+                        borderRadius: "15px",
+                        boxShadow: "0 3px 5px rgba(0,0,0,0.8)",
+                        margin: "auto",
+                        borderColor: confirmationErrors.length == 0 ? "#a8d49c" : "#990f0f",
+                        borderWidth: "1px",
+                        borderStyle: "solid",
```
Is `sx` required for a Dialog as it is for `DataGrid`, or can we do this with a styled component? Same for all the ones below where we use sx.

This is a low priority change, so probably leave as is but tell me whether we could move the styling into styled components and I can do it later.

ebinezer04 replied:
sx is not required for the Dialog or the Button so styled components can be used

Stuart-Knowles-SW commented on src/app/parcels/batch/displayComponents/BatchParcelDataGrid.tsx:
```diff
@@ -37,65 +45,277 @@ export interface BatchGridDisplayRow {
+                        >
+                            Close
+                        </Button>
+                        {confirmationErrors.length == 0 && (
+                            <Button
+                                variant="contained"
+                                onClick={handleSubmit}
+                                sx={{ margin: "1rem", minWidth: "120px" }}
+                            >
+                                Submit
+                            </Button>
```
Rather than having this conditionally rendered, could we `disable` it based on the errors?

Stuart-Knowles-SW commented on src/app/parcels/batch/displayComponents/BatchParcelDataGrid.tsx:
```diff
@@ -37,65 +45,277 @@ export interface BatchGridDisplayRow {
+                )}
+                <DialogActions sx={{ justifyContent: "center" }}>
+                    {submitErrors.length > 0 ? (
+                        <>
+                            <Button
+                                variant="contained"
+                                sx={{ margin: "1rem", minWidth: "120px" }}
+                                onClick={() => {
+                                    handleReset();
+                                    setIsSuccessDialogVisible(false);
+                                }}
+                            >
+                                Reset rows and start again
+                            </Button>
+                            <Button
+                                variant="contained"
+                                onClick={() => setIsSuccessDialogVisible(false)}
+                                sx={{ margin: "1rem", minWidth: "120px" }}
+                            >
+                                View unsubmitted rows
+                            </Button>
+                        </>
+                    ) : (
+                        <>
+                            <Button
+                                variant="contained"
+                                sx={{ margin: "1rem", minWidth: "120px" }}
+                                onClick={() => setIsSuccessDialogVisible(false)}
+                            >
+                                Add more rows
+                            </Button>
+                            <Button
+                                variant="contained"
+                                sx={{ margin: "1rem", minWidth: "120px" }}
+                                onClick={handleReturnToParcels}
+                            >
+                                Return to the Parcels Page
+                            </Button>
+                        </>
+                    )}
```
I'd ideally like these two options split out into their own components. Something like:
```
                <DialogActions sx={{ justifyContent: "center" }}>
                    {submitErrors.length > 0 ? 
                        <BatchParcelsPostSumbitErrorButtons ... />
                     : <BatchParcelsPostSumbitSuccessButtons ... />
                    )}
```

ebinezer04 replied:
Implemented this. It might be worth doing the same for the separate dialogs for confirmation and post submission.

Stuart-Knowles-SW commented on src/app/parcels/batch/submitTableData.ts:
```diff
@@ -0,0 +1,250 @@
+const batchClientToInsertRecords = (
+    client: BatchClient
+): { clientRecord: ClientDatabaseInsertRecord; familyMembers: FamilyDatabaseInsertRecord[] } => {
+    const adults: Person[] = client.adultInfo?.adults || [];
+    const children: Person[] = client.childrenInfo?.children || [];
```
same here, let's use `??` not `||`

Stuart-Knowles-SW reviewed (commented):
Few changes but overall this looks really good, really nice work on the submission and verification functions

Stuart-Knowles-SW committed:
update with new types

Stuart-Knowles-SW committed:
move batch-create into a top-level route

Stuart-Knowles-SW committed:
fix button resizing on focus

Stuart-Knowles-SW committed:
fix nav bar

Stuart-Knowles-SW committed:
Add character limit to batch creation text inputs

Stuart-Knowles-SW committed:
Redo directory structure of batch-create for clarity

Stuart-Knowles-SW committed:
Update batch-create page data

Stuart-Knowles-SW committed:
Refactor action buttons

Stuart-Knowles-SW committed:
Allow deleting rows and fix batch creation scrolling

Stuart-Knowles-SW committed:
fix test

Stuart-Knowles-SW commented:
I've ended up bundling a lot of stuff into this PR which didn't really belong here, but for the sake of expediency we're going to push ahead with this
